src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -0,0 +1,243 @@
+    /// <see cref="DisconnectedCircuits"/> should ensure we no longer have to concern ourselves with entry expiration.
+    ///
+    /// Knowing when a client disconnected is not an exact science. There's a fair possiblity that a client may reconnect before the server realizes.
+    /// Consequently, we have to account for reconnects and disconnects occuring simultaneously as well as appearing out of order.
+    /// To manage this, we use a critical section to manage all state transitions.
+    /// </remarks>
+    internal class CircuitRegistry
+    {
+        private readonly object CircuitRegistryLock = new object();
+        private readonly ComponentsServerOptions _options;
+        private readonly ILogger _logger;
+        private readonly PostEvictionCallbackRegistration _postEvictionCallback;
+
+        public CircuitRegistry(
+            IOptions<ComponentsServerOptions> options,
+            ILogger<CircuitRegistry> logger)
+        {
+            _options = options.Value;
+            _logger = logger;
+
+            ConnectedCircuits = new ConcurrentDictionary<string, CircuitHost>(StringComparer.Ordinal);
+
+            DisconnectedCircuits = new MemoryCache(new MemoryCacheOptions
+            {
+                SizeLimit = _options.MaxRetainedDisconnectedCircuits,
+            });
+
+            _postEvictionCallback = new PostEvictionCallbackRegistration
+            {
+                EvictionCallback = OnEntryEvicted,
+            };
+        }
+
+        internal ConcurrentDictionary<string, CircuitHost> ConnectedCircuits { get; }
+
+        internal MemoryCache DisconnectedCircuits { get; }
+
+        /// <summary>
+        /// Registers an active <see cref="CircuitHost"/> with the register.
+        /// </summary>
+        public void Register(CircuitHost circuitHost)
+        {
+            if (!ConnectedCircuits.TryAdd(circuitHost.CircuitId, circuitHost))
+            {
+                throw new ArgumentException($"Circuit with identity {circuitHost.CircuitId} is already registered.");
+            }
+        }
+
+        public virtual Task DisconnectAsync(CircuitHost circuitHost, string connectionId)
+        {
+            bool disconnected;
+            lock (CircuitRegistryLock)
+            {
+                disconnected = DisconnectCore(circuitHost, connectionId);
+            }
+
+            if (disconnected)
+            {
+                // DisconnectCore may fail to disconnect the circuit if it was previously marked inactive or
+                // has been transfered to a new connection. Do not invoke the circuit handlers in this instance.
+
+                // CircuitHandler events are invoked outside the critical section. This may result in simultaneous
+                // execution of a disconnect and connect events, however we make no guarantees of the order in which they execute.
+                return circuitHost.OnConnectionDownAsync();
```

Comment:
I'm not entirely certain that this is the best behaviour. Imagine that you're using the events to serialize state to redis - how would you rationalize that? Are we basically going to say that this is the best we can do because you have no guarantees in a multi-server environment?

Reply:
We could try and guarantee ordering, but there's a fair chance that the server does not know the client is down. I don't think connection up \ down is a very good event to decide when to serialize state.

Reply:
I think it's fair to see that connection up-down isn't the best event to serialize state then. I think it's a mind-boggling decision to try and make extensibility authors do their own per-circuit synchronization. They don't have the tools to do that.

So the following would make sense:

1. Provide connection up/down notifications. Guarantee ordering so that you don't see the same event twice in a row. Guarantee that events are not concurrent.

2. Provide connection up/down/changed notifications. Guarantee that each up is eventually paired with a down, and you never see an up+up or down+down pair. Guarantee that events are not concurrent.

Reply:
> 1. Provide connection up/down notifications. Guarantee ordering so that you don't see the same event twice in a row. 

I guess we could simulate a reconnect i.e. Connected -> Connected would result in connection down + connection up. 

> Guarantee that events are not concurrent.

`SemaphoreSlim` it is again. Oh well

Reply:
Well, I think we can have a discussion before you do a million works to rewrite this code. I just think that this is fundamentally too complicated to be useful rn to a user. 

Do we want to start over and think about what scenarios matter first, and then make a decision? We don't have to block on this part.